pkg/download/downloader.go

```diff
@@ -92,6 +107,38 @@ func extractZIP(targetDir string, read io.ReaderAt, size int64) error {
 	return nil
 }
 
+func isSymlinkAllowed(targetDir, oldname, newpath string) error {
+	// no symlinks to absolute paths
+	if filepath.IsAbs(oldname) {
+		glog.V(4).Infof("invalid absolute symlink %v -> %v\n", oldname, newpath)
+		return errors.New("invalid symlink referencing an absolute path in tar")
+	}
+
+	oldpath := filepath.Join(filepath.Dir(newpath), filepath.Clean(oldname))
+	fmt.Printf("LINK %s %s\n", oldpath, newpath)
+
+	if _, isNonEscaping := pathutil.IsSubPath(targetDir, oldpath); !isNonEscaping {
+		glog.V(4).Infof("invalid escaping symlink %v -> %v\n", oldname, newpath)
+		return errors.New("invalid symlink referencing a parent path in tar")
+	}
+
+	return nil
+}
+
+// make a symlink from oldname to newname within the given enclosing targetDir
+// enforces two symlink policies
+// 1) no symlinks to absolute paths (oldname must not be an absolute path)
+// 2) no symlinks that escape targetDir
+func symlinkIfAllowed(targetDir, oldname, newpath string) error {
+	if err := isSymlinkAllowed(targetDir, oldname, newpath); err != nil {
+		return err
+	}
+	if err := os.Symlink(oldname, newpath); err != nil {
+		return errors.Wrap(err, "failed to create symlink from tar")
```

Comment:
nit: you can actually directly return this `errors.Wrap` result as when err is nil, it will return nil.